### PR TITLE
Add loader to thread list

### LINKED ISSUE #5562: No loading spinner is shown while threads list is loading
### Steps to reproduce

1. Go to a room with lots of threads (e.g. Threads internal)
2. View the thread list on a fresh login

#### What did you expect?

To either see the thread list or a loading spinner telling me whats up

#### What happened instead?

The threads list looked empty for about 20 seconds, I assumed it was broken somehow

### Your phone model

Galaxy S20

### Operating system version

Android 11

### Application version and app store

1.4.6-dev [40104060] (G-b6874) feature/aris/thread_live_thread_list

### Homeserver

matrix.org

### Will you send logs?

No

### PR DESCRIPTION
**Pull request by ariskotsomitopoulos**
## Type of change

- [ ] Feature
- [x] Bugfix
- [ ] Technical
- [ ] Other :

## Content

Add loader when user open/load the thread list

## Motivation and context

Especially on large rooms and on initial launch of the thread list, we should show an indicator to the user about the progress

## Screenshots / GIFs
|Before|After|
|---|---|
|Screenshot_20220330-121034_Element dbg|Screenshot_20220330-121026_Element dbg|

Closes #5562

### CODE REVIEW

File: vector/src/main/java/im/vector/app/features/home/room/threads/list/viewmodel/ThreadListViewModel.kt [hunk 1]
```diff
@@ -113,7 +113,15 @@ class ThreadListViewModel @AssistedInject constructor(@Assisted val initialState
 
     private fun fetchThreadList() {
         viewModelScope.launch {
+            isLoading(true)
             room?.fetchThreadSummaries()
+            isLoading(false)
+        }
+    }
+
+    private fun isLoading(show: Boolean) {
```
Comment (onurays) [inline vector/src/main/java/im/vector/app/features/home/room/threads/list/viewmodel/ThreadListViewModel.kt:122]: I think `private fun setLoading(isLoading: Boolean)` is more convenient.

File: vector/src/main/res/layout/fragment_thread_list.xml [hunk 2]
```diff
@@ -37,6 +37,21 @@
...
+
+        <ProgressBar
+            android:id="@+id/threadListProgressBar"
+            style="@style/Widget.Vector.ProgressBar.Horizontal"
+            android:layout_width="0dp"
+            android:layout_height="wrap_content"
+            android:layout_gravity="center_vertical"
+            android:indeterminate="true"
+            android:visibility="gone"
+            android:layout_marginTop="-6dp"
```
Comment (onurays) [inline vector/src/main/res/layout/fragment_thread_list.xml:49]: I see why we need this but we have a lint rule for negative margins. Is it a problem to show the tiny progressbar on the first message? If it is we need `tools:ignore="NegativeMargin"` at least.

Comment (onurays) [reply vector/src/main/res/layout/fragment_thread_list.xml:49]: Or `app:layout_constraintBottom_toBottomOf="@id/threadListAppBarLayout"` solves the problem without negative margin?

**Review by onurays — CHANGES_REQUESTED** [review]
Nice improvement! I have some minor remarks.

File: vector/src/main/res/layout/fragment_thread_list.xml (hunk 2 above)
Comment (ariskotsomitopoulos) [reply vector/src/main/res/layout/fragment_thread_list.xml:49]: For that known problem I used to wrap it around a fragment, that eliminates the problem, but this adds an extra layer and decrease code readability.  So I prefer to use `-6dp` as an exception in my projects, never had an issue with that.

Regarding `app:layout_constraintBottom_toBottomOf="@id/threadListAppBarLayout"` will not have any difference while the padding is either on top or on the bottom.

P.S. On Jetpack Compose they fixed that issues the horizontal progressBar ! 🥇

File: vector/src/main/java/im/vector/app/features/home/room/threads/list/viewmodel/ThreadListViewModel.kt (hunk 1 above)
Comment (ariskotsomitopoulos) [reply vector/src/main/java/im/vector/app/features/home/room/threads/list/viewmodel/ThreadListViewModel.kt]: Refactored!

**Review by onurays — APPROVED** [review]
Thank you for the update.

File: vector/src/main/java/im/vector/app/features/home/room/threads/list/views/ThreadListFragment.kt [hunk 3]
```diff
@@ -104,6 +104,11 @@ class ThreadListFragment @Inject constructor(
     override fun invalidate() = withState(threadListViewModel) { state ->
         renderEmptyStateIfNeeded(state)
         threadListController.update(state)
+        renderLoaderIfNeeded(state)
+    }
+
+    private fun renderLoaderIfNeeded(state: ThreadListViewState) {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/home/room/threads/list/views/ThreadListFragment.kt:110]: `IfNeeded` suffix is not necessary here I guess.

File: vector/src/main/res/layout/fragment_thread_list.xml (hunk 2 above)
Comment (bmarty) [reply vector/src/main/res/layout/fragment_thread_list.xml:50]: You could have reused what we are doing in the room list / room timeline: https://github.com/vector-im/element-android/blob/main/vector/src/main/res/layout/view_sync_state.xml#L38

**Review by bmarty — COMMENTED** [review]
Some post merge remarks :)

File: vector/src/main/java/im/vector/app/features/home/room/threads/list/views/ThreadListFragment.kt (hunk 3 above)
Comment (ariskotsomitopoulos) [reply vector/src/main/java/im/vector/app/features/home/room/threads/list/views/ThreadListFragment.kt:110]: Correct, its there from a previous approach!

File: vector/src/main/res/layout/fragment_thread_list.xml (hunk 2 above)
Comment (ariskotsomitopoulos) [reply vector/src/main/res/layout/fragment_thread_list.xml:50]: Yes I saw it but, nesting affects performance and makes code less readable, I cannot see any benefit from that. Do you have a strong point?

**Pull request merged**